lazy_dataset/core.py

```diff
@@ -2730,6 +2784,132 @@ def __str__(self):
             return super().__str__()
 
 
+class _DiskCacheWrapper:
+    """
+    Wraps a `diskcache.Cache` and takes care of cleaning up when destroyed.
+    The wrapper is required to support sharing the cache among copies of the
+    dataset.
+    """
+    def __init__(self, cache_dir, reuse, clear):
+        self.clear = clear
+
+        import diskcache
+        if cache_dir is not None and Path(cache_dir).is_dir():
+            if reuse:
+                print(f'Cache dir "{cache_dir}" already exists. Re-using '
+                      f'stored data.')
+            else:
+                raise RuntimeError(
+                    f'Cache dir "{cache_dir}" exists! Either remove it '
+                    f'or set reuse=True.'
+                )
+        self.cache = diskcache.Cache(cache_dir)
+
+    def __del__(self):
+        # This gets called when all references to the cache wrapper are
+        # dropped and the program is still running. This includes normal
+        # termination and keyboard interrupt, but no other signals like
+        # SIGTERM or SIGKILL. Some signals sometimes work if they are handled
+        # within python.
+        self.cache.close()
+        if self.clear:
+            if Path(self.cache.directory).exists():
+                import shutil
+                shutil.rmtree(self.cache.directory)
+
+
+class DiskCacheDataset(Dataset):
+    def __init__(self, input_dataset, cache_dir=None, reuse=True, clear=True,
+                 *, _cache=None):
+        # We have the same assumptions as CacheDataset
+        self.input_dataset = input_dataset
+        assert self.input_dataset.indexable
+        if _cache is None:
+            self.cache = _DiskCacheWrapper(cache_dir, reuse, clear)
+        else:
+            self.cache = _cache
+
+    @property
+    def indexable(self) -> bool:
+        return self.input_dataset.indexable
+
+    @property
+    def ordered(self) -> bool:
+        return self.input_dataset.ordered
+
+    def keys(self) -> list:
+        return self.input_dataset.keys()
+
+    def __iter__(self):
+        for i in range(len(self)):
+            yield self[i]
+
+    def __len__(self):
+        return len(self.input_dataset)
+
+    def __getitem__(self, item):
+        if isinstance(item, str):
+            item = self.keys().index(item)
+
+        if isinstance(item, numbers.Integral):
+            # diskcache already provides new copy of the item on every read,
+            # so we don't have to think about immutability of examples
+            if item not in self.cache.cache:
+
+                import shutil
+                import humanfriendly
+                diskusage = shutil.disk_usage(self.cache.cache.directory)
```

Comment:
Do you know, how expensive this operation is?

Reply:
No, I don't know. But I thought in cases where you want to use this type of cache, a little more delay is not crucial. But I can check and remove it or reduce the frequency it is called

Reply:
Thanks

Reply:
```python
%%timeit
shutil.disk_usage('/tmp')
```
gives `1.75 µs ± 97.4 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)`, so the overhead seems to be small

Reply:
That is much faster than I expected. I think there is a caching involved. I tried it with a single call and got a value around `20 µs`. Nonetheless, the overhead is small enough.